ProjFS.Mac/PrjFSKext/AssociativeArrayTests/AssociativeArrayTests.mm

```diff
@@ -0,0 +1,135 @@
+#import <XCTest/XCTest.h>
```

Comment:
This file looks like it's Objective C perhaps? Can you give us some context on how this is all working?

Reply:
Yeah, the `XCTest` framework is designed for Objective-C and Swift. You can mix Objective-C and C++ in the combined offshoot that is Objective-C++ - this is what the `.mm` suffix indicates, `.m` would be Objective-C. (c.f. `.c` for C, and `.cc` for C++)

All methods whose name begins with "test" in any subclasses of XCTestCase are run as tests, success and failure determined by the outcomes of any `XCTestAssert` macros. While entirely OK for testing C++ code via the detour to Objective-C++, there certainly remains a certain syntactic disconnect, and there are some oddities lurking in the assert macros. (such as use of `self`, which makes placing asserts in C++ functions a little trickier, for example in the [test `Memory_Free()`](https://github.com/Microsoft/VFSForGit/pull/500/files#diff-3971b7af86d2ce078fb9151e50d21928R48).)

If the Scheme is configured correctly, you can hit the "Test" button (click and hold "Run" in Xcode, then select "Test" from the dropdown) and the tests will run straight in Xcode, failing tests are shown in the warning/error UI, and success/failure is indicated on the source code line where each test is defined. `xcodebuild test` runs the tests from the command line.

As we've avoided bringing in any external dependencies so far, I thought I'd prototype it with XCTest and find out what you all think of the overall approach. (And whether we even want to start unit testing bits of the kext code outside the kext itself.) There's certainly no shortage of alternatives. I can see that [Visual Studio ships with 4 different unit test frameworks](https://docs.microsoft.com/en-us/visualstudio/test/writing-unit-tests-for-c-cpp?view=vs-2017) so that's probably a good starting point:

 * I can't seem to find a way to obtain the "Microsoft Unit Testing Framework for C++" independently from Visual Studio, so I'm guessing that might be a nonstarter on macOS.
 * CTest only really makes sense when using CMake for the build; we're not currently using that on VFS For Git on macOS, so it's probably not a good choice unless we also change our build system.
 * That leaves Google Test and Boost.Test; both work on macOS. The lowest-friction way to get them running is probably via homebrew or MacPorts, but that's probably not so great for the test farm. Google Test is very easy to fetch & build on a clean macOS system with only Xcode installed.

Anyway, this probably is all irrelevant if we don't go with the templated data structure approach, unit testing the intertwined data structure/business logic is probably going to be more trouble than it's worth.

Reply:
My opinion here is to keep using XCTest since we will only ever be using this code on MacOS. If it needed to be cross-plat, Google Test sounds like a good choice but I'd rather us use the native solution that our environment (Xcode) provides rather than add a new large dependency. 

We could easily call ```xcodebuild test``` or ```xctest -Test``` from our Build and UT script.

I'm a little biased though as I've used XCTest before and think it works well for unit testing for native MacOS code. I could be persuaded though :)

Reply:
When I get a chance, I'll see how easy it is to wrap the most awkward of Objective-C bits in something more palatable to C++ developers. The reliance on `self` is my main grievance with XCTest, which is what necessitates the contortions for carrying around the test object:

```
static AssociativeArrayTests* s_test;
…
void Memory_Free(void* memory, uint32_t size)
{
    AssociativeArrayTests* self = s_test;
```

There's got to be a better way to do that in C++, but there are more pressing concerns at the moment. :-)